File: OJ/leetcode/cpp/40_combination_sum_ii.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void dfs(int start, vector<int>& stack, vector<vector<int>>& ans, 
         const vector<int>& candidates, int target) {
    if (target <= 0) {
        if (target == 0) {
            ans.push_back(stack);
        }
        return;
    }

    for (int i = start; i < candidates.size(); ++i) {
        if (i != start && candidates[i] == candidates[i - 1]) {
            continue;
        }
        stack.push_back(candidates[i]);
        dfs(i + 1, stack, ans, candidates, target - candidates[i]);
        stack.pop_back();
    }
}

vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    sort(candidates.begin(), candidates.end());
    vector<vector<int>> ans;
    vector<int> stack;
    dfs(0, stack, ans, candidates, target);
    return ans;
}

int main() {
    vector<int> candidates = {2, 5, 2, 1, 2};
    vector<vector<int>> ans = combinationSum2(candidates, 5);
    for (vector<int> vec : ans) {
        for (int i : vec) {
            cout << i << " ";
        }
        cout << "\n";
    }
}

File: competitive_programming/leetcode/python/40_combination_sum_ii.py
# TLE
def combinationSum(candidates, target):
  index_ans = []
  candidates.sort()
  count = target // candidates[0]
  n = len(candidates)
  tree, ans = [[]], []

  while count > 0:
    tmp = []
    for arr in tree:
      start = 0 if len(arr) == 0 else arr[-1] + 1
      for i in range(start, n):
        sum_ = sum([candidates[j] for j in arr]) + candidates[i]
        if sum_ < target:
          tmp.append(arr + [i])
        elif sum_ == target:
          index_ans.append(arr + [i])
    tree = tmp 
    count -= 1

  hash_table = set()
  for arr in index_ans:
    s = "".join([str(candidates[i]) for i in arr])
    if s not in hash_table:
      ans.append([candidates[i] for i in arr])
      hash_table.add(s)

  return ans

def dfsSolution(candidates, target):
  def dfs(candidates, target, start, stack, ans):
    if target == 0:
      ans.append(stack[:])
      return
    
    for i in range(start, len(candidates)):
      if candidates[i] > target:
        return
      # prune - omit equivalant elements in same depth
      if i > start and candidates[i] == candidates[i - 1]: 
        continue
      stack.append(candidates[i])
      dfs(candidates, target - candidates[i], i + 1, stack, ans)
      stack.pop()

  candidates.sort()
  ans = []
  dfs(candidates, target, 0, [], ans)
  return ans

print(dfsSolution([10, 1, 2, 7, 6, 1, 5], 8))
print(dfsSolution([2, 5, 2, 1, 2], 5))
print(dfsSolution([1, 1], 2))
print(dfsSolution(
    [1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1], 27
    )
)